### PR TITLE
Speed up Reading RetentionLeases from the Wire

### PR DESCRIPTION
**Pull request by original-brownbear**
This is showing up quite hot for stats (per shard stats) on the transport threads.
No need to mess with the stream and to allocate an intermediary list here when we
can read directly into an efficient map.

relates #77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java [hunk 1]
```diff
@@ -149,7 +167,22 @@ public RetentionLeases(final long primaryTerm, final long version, final Collect
...
+                final RetentionLease lease = new RetentionLease(in);
+                leases = Map.of(lease.id(), lease);
+            }
+            default -> {
+                final Map<String, RetentionLease> m = Maps.newLinkedHashMapWithExpectedSize(leasesCount);
+                for (int i = 0; i < leasesCount; i++) {
+                    final RetentionLease lease = new RetentionLease(in);
+                    m.put(lease.id(), lease);
+                }
+                leases = Collections.unmodifiableMap(m);
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java:183]: Consider using `StreamInput::readMapValues`. May be that one could be improved to have a special case for size=1

File: server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java [hunk 2]
```diff
@@ -136,7 +136,25 @@ public RetentionLeases(final long primaryTerm, final long version, final Collect
...
+            case 0 -> this.leases = Map.of();
+            case 1 -> {
+                final RetentionLease lease = leases.get(0);
+                this.leases = Map.of(lease.id(), lease);
+            }
+            default -> {
+                // use a linked hash map to preserve order
+                final LinkedHashMap<String, RetentionLease> map = Maps.newLinkedHashMapWithExpectedSize(leasesCount);
+                for (RetentionLease lease : leases) {
+                    map.merge(lease.id(), lease, (left, right) -> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java:150]: Any reason we don't just use `Map#put` here and throw the `ISE` if it doesn't return `null`?

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java]: Right ... I guess I just mindlessly refactored from the existing code and liked how short this was :) put and throw it is, likely faster this way.

File: server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java [hunk 3]
```diff
@@ -149,7 +167,22 @@ public RetentionLeases(final long primaryTerm, final long version, final Collect
...
+                leases = Map.of(lease.id(), lease);
+            }
+            default -> {
+                final Map<String, RetentionLease> m = Maps.newLinkedHashMapWithExpectedSize(leasesCount);
+                for (int i = 0; i < leasesCount; i++) {
+                    final RetentionLease lease = new RetentionLease(in);
+                    m.put(lease.id(), lease);
+                }
+                leases = Collections.unmodifiableMap(m);
+            }
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/seqno/RetentionLeases.java:189]: Makes sense, lets do that kind of thing in a follow-up? I think I'd like to benchmark this a little before blowing up the size of these methods with an `n==1` special case first :) (though it's likely a win in almost all read-world cases)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1  (unrelated and known)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1  (unrelated and known)

**Comment by original-brownbear** [comment]
Thanks both! I think this is trivial enough for one review to suffice :) I'll create a follow-up for the suggested wider refactoring :)

**Pull request merged**